django/db/backends/postgresql/base.py

```diff
@@ -221,35 +254,68 @@ def get_new_connection(self, conn_params):
         #   will set connection.isolation_level to ISOLATION_LEVEL_AUTOCOMMIT.
         options = self.settings_dict["OPTIONS"]
         try:
-            self.isolation_level = options["isolation_level"]
+            isolevel = options["isolation_level"]
         except KeyError:
-            self.isolation_level = connection.isolation_level
+            self.isolation_level = IsolationLevel.READ_COMMITTED
         else:
-            # Set the isolation level to the value from OPTIONS.
-            if self.isolation_level != connection.isolation_level:
-                connection.set_session(isolation_level=self.isolation_level)
-        # Register dummy loads() to avoid a round trip from psycopg2's decode
-        # to json.dumps() to json.loads(), when using a custom decoder in
-        # JSONField.
-        psycopg2.extras.register_default_jsonb(
-            conn_or_curs=connection, loads=lambda x: x
-        )
+            try:
+                self.isolation_level = IsolationLevel(isolevel)
+            except ValueError:
+                raise ImproperlyConfigured(
+                    "bad isolation_level: %s. Choose one of the "
+                    "'psycopg.IsolationLevel' values" % (options["isolation_level"],)
+                )
+            connection.isolation_level = self.isolation_level
+
+        connection.cursor_factory = Cursor
+
         return connection
 
     def ensure_timezone(self):
         if self.connection is None:
             return False
-        conn_timezone_name = self.connection.get_parameter_status("TimeZone")
+
+        conn_timezone_name = self.connection.info.parameter_status("TimeZone")
         timezone_name = self.timezone_name
         if timezone_name and conn_timezone_name != timezone_name:
             with self.connection.cursor() as cursor:
                 cursor.execute(self.ops.set_time_zone_sql(), [timezone_name])
             return True
         return False
 
+    def get_adapters_template(self):
+        key = settings.USE_TZ
```

Comment:
rather than custom caching with a dict, this might be clearer with a module-level function using `@lru_cache(maxsize=2)`, with the current value of `USE_TZ` as the only argument. It would save some lines and clarify it's a cache.

Reply:
Adjusted for now, that said it feels like a hack :D Django's settings don't change at runtime so we should simply cache one object and reset it via the test signals if that changes?

Reply:
Hmmm, that would add some overhead for a bunch of individual tests marked with `@override_settings(USE_TZ=...)`, since they'd repeatedly construct the different adapter templates,

Reply:
Yeah, I guess I'll leave it as is (well that is with a properly working cache ;))